### PR TITLE
spotifyd: generate TOML config via formats

### PR DESCRIPTION
**Pull request by kmein**
With version 0.3.0, spotifyd switched to a new TOML config file format (as discussed in [this issue](https://github.com/Spotifyd/spotifyd/issues/765)). I took the opportunity to use `pkgs.formats.toml` to generate such a config and bring the module's options in line with the policy of having a `settings` option (as opposed to a `config` option), while keeping the module backwards-compatible.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by anderslundstedt** [comment]
With nothing but `services.spotifyd.enable = true`, I get:

```
...
building Nix...
building the system configuration...
error:
Failed assertions:
- Using the stringly typed .config attribute is discouraged. Use the TOML typed .settings attribute instead.
(use '--show-trace' to show detailed location information)
```

File: nixos/modules/services/audio/spotifyd.nix [hunk 1]
```diff
@@ -14,6 +15,15 @@ in
...
         type = types.lines;
+        description = ''
+          (Deprecated) Configuration for Spotifyd. For syntax and directives, see
+          <link xlink:href="https://github.com/Spotifyd/spotifyd#Configuration"/>.
+        '';
+      };
+
+      settings = mkOption {
+        default = {};
+        type = toml.type;
```
Comment (anderslundstedt) [inline nixos/modules/services/audio/spotifyd.nix:26]: Perhaps an example value would be nice? I had to guess that the format was

```
services.spotifyd.settings = {
  public = {
    bitrate = 320;
  };
};
```

File: nixos/modules/services/audio/spotifyd.nix [hunk 2]
```diff
@@ -23,6 +33,13 @@ in
   };
 
   config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.config == "" || cfg.settings == {};
+        message = "Using the stringly typed .config attribute is discouraged. Use the TOML typed .settings attribute instead.";
+      }
+    ];
+
```
Comment (anderslundstedt) [inline nixos/modules/services/audio/spotifyd.nix:42]: Should the error message not be something like "at most one of the .config attribute and the .settings attribute may be set"?

**Review by anderslundstedt — COMMENTED** [review]
Builds fine now. I have never used the `.config` option but both that and the new `.settings` option seem to work fine.

I guess a warning should be shown if the `.config` option is set?

File: nixos/modules/services/audio/spotifyd.nix [hunk 3]
```diff
@@ -14,6 +18,16 @@ in
...
+        description = ''
+          (Deprecated) Configuration for Spotifyd. For syntax and directives, see
+          <link xlink:href="https://github.com/Spotifyd/spotifyd#Configuration"/>.
+        '';
+      };
+
+      settings = mkOption {
+        default = {};
+        type = toml.type;
+        example = { public.bitrate = 320; };
```
Comment (anderslundstedt) [inline nixos/modules/services/audio/spotifyd.nix:30]: My bad, `public` should be `global`.

```suggestion
        example = { global.bitrate = 320; };
```

File: nixos/modules/services/audio/spotifyd.nix [hunk 4]
```diff
@@ -4,7 +4,11 @@ with lib;
 
 let
   cfg = config.services.spotifyd;
-  spotifydConf = pkgs.writeText "spotifyd.conf" cfg.config;
+  toml = pkgs.formats.toml {};
+  spotifydConf =
+    if cfg.settings != {}
+    then toml.generate "spotify.conf" cfg.settings
+    else lib.trace "Using the stringly typed .config attribute is discouraged. Use the TOML typed .settings attribute instead." (pkgs.writeText "spotifyd.conf" cfg.config);
```
Comment (anderslundstedt) [inline nixos/modules/services/audio/spotifyd.nix:11]: The warning is now shown when only `services.spotifyd.enable = true;` is set (I do not think it was before, or perhaps I missed it).

**Review by anderslundstedt — CHANGES_REQUESTED** [review]
Minor things remaining. Everything seems to work fine.

File: nixos/modules/services/audio/spotifyd.nix [hunk 5]
```diff
@@ -4,7 +4,15 @@ with lib;
 
 let
   cfg = config.services.spotifyd;
-  spotifydConf = pkgs.writeText "spotifyd.conf" cfg.config;
+  toml = pkgs.formats.toml {};
+  warnConfig =
+    if cfg.config != ""
+    then lib.trace "Using the stringly typed .config attribute is discouraged. Use the TOML typed .settings attribute instead."
+    else identity;
```
Comment (anderslundstedt) [inline nixos/modules/services/audio/spotifyd.nix:11]: I get

```
error: undefined variable 'identity' at /home/anders/nixpkgs/nixos/modules/services/audio/spotifyd.nix:11:10
```

Did you mean the below? Everything seems to work correctly for me with the below change.

```suggestion
    else id;
```

**Review by anderslundstedt — CHANGES_REQUESTED** [review]

File: nixos/modules/services/audio/spotifyd.nix (hunk 5 above)
Comment (kmein) [reply nixos/modules/services/audio/spotifyd.nix:11]: Yes, of course, sorry.

**Review by anderslundstedt — APPROVED** [review]

**Pull request merged**